micro_http/src/server.rs

```diff
@@ -0,0 +1,551 @@
+                        Err(error) => return Err(error),
+                        Ok(()) => {}
+                    };
+                }
+                token => {
+                    self.serve_connection(
+                        token,
+                        e.events,
+                        &mut streams_to_be_removed,
+                        &mut parsed_requests,
+                    )?;
+                }
+            }
+        }
+        for token in streams_to_be_removed {
+            self.del_stream(token);
+        }
+
+        Ok(parsed_requests)
+    }
+
+    /// Returns the file descriptor of the server's internal `epoll` structure.
+    pub fn get_epoll_fd(&self) -> RawFd {
+        self.epoll_fd
+    }
+
+    /// Enqueues the provided responses in the outgoing connection.
+    pub fn enqueue_responses(&mut self, responses: Vec<(Token, Response)>) -> Result<()> {
+        for (token, response) in responses {
+            if let Some(file_streamed_connection) = self.connections.get_mut(&token) {
+                file_streamed_connection
+                    .connection
+                    .enqueue_response(response);
+
+                epoll::ctl(
+                    self.epoll_fd,
+                    epoll::ControlOptions::EPOLL_CTL_MOD,
+                    file_streamed_connection.as_raw_fd(),
+                    epoll::Event::new(epoll::Events::EPOLLOUT, token as u64),
+                )
+                .map_err(|_| ServerError::IOError)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    // Accepts a new incoming connection and adds it to the `epoll` notification structure.
+    fn handle_new_connection(&mut self) -> Result<()> {
+        let (mut new_stream, _new_addr) = self.socket.accept().expect("Socket accept failed.");
+        if self.token_queue.is_empty() {
+            // If we want a replacement policy for connections
+            // this is where we will have it.
+            match new_stream.write(SERVER_FULL_ERROR_MESSAGE) {
+                _ => return Err(ServerError::ServerFull),
+            }
+        }
+
+        let stream_fd = new_stream.as_raw_fd();
+        new_stream
+            .set_nonblocking(true)
+            .map_err(|_| ServerError::IOError)?;
+        // Safe because we checked that there are available tokens.
+        let token = self.token_queue.pop_front().unwrap();
+        epoll::ctl(
+            self.epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            new_stream.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, token as u64),
+        )
+        .expect("Cannot accept new connection");
+        let new_conn = HttpConnection::new(new_stream);
+        self.connections
+            .insert(token, ConnectionWithFd::new(new_conn, stream_fd));
+        Ok(())
+    }
+
+    // Serves a connection that has been notified by the `epoll` structure.
+    fn serve_connection(
```

Comment:
This functionality should happen inside a `Connection` object - `conn.process()` like described above.